### ISSUE TITLE
Pinch zoom crashes the touch control in slow/old mobile browsers (especially iOS)

### ISSUE THREAD

**Issue by t-muko** [openlayers/openlayers#11465]
**Describe the bug**
Pinch zooming a page when the renderer/browser gets busy, puts the map in a state where zooming no longer works and panning is woking "in half speed". May be the same issue as with other touch problems where "an invisible finger" is stuck "on" and single finger pan is actually zoom with very wide finger distance.

Symptoms are somewhat similar to #11251.

**To Reproduce**
1. Using iOS 12 device zoom in and out using the attached test code
2. After a while the zooming gestures no longer work
3. The only way to recover is to reload the page

iOS crash happens both in Safari and Chrome, so the issue might be renderer related.

**Expected behavior**
Two finger zooming does not crash the map.

Sample code:
index.html:
```
<!DOCTYPE html>
<html lang="en" style="height: 100%;">
  <head>
    <title>Simple Map</title>
    <!-- Pointer events polyfill for old browsers, see https://caniuse.com/#feat=pointer -->
    <script src="https://unpkg.com/elm-pep"></script>
    <style>
      .map {
        width: 100%;
        height: 100%;
      }
    </style>
  </head>
  <body style="height: 100%;">
    <div id="map" class="map"></div>
    <script src="main.js"></script>
  </body>
</html>
```

main.js
```
import 'ol/ol.css';
import Map from 'ol/Map';
import OSM from 'ol/source/OSM';
import TileLayer from 'ol/layer/Tile';
import View from 'ol/View';
import { Circle as CircleStyle, Fill, Stroke, Style, Text } from 'ol/style';
import VectorTileLayer from 'ol/layer/VectorTile';
import VectorTileSource from 'ol/source/VectorTile';
import OSMXML from 'ol/format/OSMXML';
import { transformExtent } from 'ol/proj';
import { defaults as defaultControls, ScaleLine, ZoomSlider } from 'ol/control';
import { fromLonLat } from 'ol/proj';



const overpass_server = "https://overpass.kumi.systems/api/interpreter"



const nameStyle = new Style({
    text: new Text({
      font: 'bold 20px "Open Sans", "Arial Unicode MS", "sans-serif"',
      placement: 'line',
      fill: new Fill({
        color: 'white'
      }),
      stroke: new Stroke({ color: 'grey', width: 3 })
    })
  });

var streetNamesOSM = new VectorTileSource({
    format: new OSMXML(),
    attributions: "-- Streetnames: OpenStreetMap",
    url: overpass_server + '/{z}/{x}/{-y}',
    maxZoom: 13,
    minZoom: 13,
    tileLoadFunction: function (tile, url) {
        console.debug("kadunnimet url " + url)

        tile.setLoader(function (extent, resolution, projection) {
            var epsg4326Extent = transformExtent(extent, projection, 'EPSG:4326');
            var query = '[bbox: ' +
                epsg4326Extent[1] + ',' + epsg4326Extent[0] + ',' +
                epsg4326Extent[3] + ',' + epsg4326Extent[2] +
                '];(way[highway="residential"]["name"];way[highway="primary"]["name"];way[highway="secondary"]["name"];way[highway="tertiary"]["name"];way[highway="unclassified"]["name"];);(._;>;);out;';

            fetch(overpass_server + '?data=' + encodeURI(query))
                .then(
                    function (response) {
                        if (response.status !== 200) {
                            console.log('Looks like there was a problem getting OSM features. Status Code: ' +
                                response.status);
                            return;
                        }

                        // Examine the text in the response
                        response.text().then(function (data) {
                            //console.log(data);
                            var features = new OSMXML().readFeatures(data, {
                                featureProjection: map.getView().getProjection()
                            });
                            tile.setFeatures(features);
                        });
                    }
                )
                .catch(function (err) {
                    console.log('Fetch Error :-S', err);
                });
        }
        )
    }
});

var nimiOSMLayer = new VectorTileLayer({
    name: "OSMstreetNames",
    source: streetNamesOSM,
    minZoom: 15.6,
    renderMode: 'hybrid',
    renderOrder: null,
    renderBuffer: 100,
    style: function (feature) {
        if (['residential', 'primary', 'secondary', 'tertiary', 'unclassified'].includes(feature.get('highway'))) {
            nameStyle.getText().setText(feature.get('name'));
            return nameStyle;
        }
        else {
            return null
        }
    }
});

var map = new Map({
    layers: [
        new TileLayer({
            source: new OSM(),
        }), nimiOSMLayer],
    target: 'map',
    view: new View({
        center: fromLonLat([-0.12, 51.50]),
        zoom: 16,
    }),
    controls: defaultControls({ zoomOptions: { delta: 0.5 } })
});
```

package.json
```
{
    "name": "simple",
    "dependencies": {
      "ol": "6.4.3"
    },
    "devDependencies": {
      "parcel": "1.11.0"
    },
    "scripts": {
      "start": "parcel index.html --host 0.0.0.0",
      "build": "parcel build --experimental-scope-hoisting --public-url . index.html"
    }
  }
```

**Comment by t-muko** [comment]
Two finger rotate stops working too when the map gets into this state.

**Comment by ahocevar** [comment]
Can you share a codesandbox with your code? Also, are you able to test on iOS 13? Just to rule out that it is related to the pointer events polyfill?

**Comment by t-muko** [comment]
I need to learn to use codesandbox first....

I have briefly tested with iOS13 and it seems to work. However, as the symptoms are very similar to other issues I have experiensed also with Android and other browsers, the pointer events might not be the one to blame here. Can't be sure, though.

**Comment by t-muko** [comment]
Codesandbox: [https://codesandbox.io/s/ancient-breeze-b1b1e?file=/main.js](https://codesandbox.io/s/ancient-breeze-b1b1e?file=/main.js)
Test-URL: [https://b1b1e.csb.app/](https://b1b1e.csb.app/)

**Comment by ahocevar** [comment]
> I need to learn to use codesandbox first....

It looks like the code you pasted is from a modified example. If that's the case, just click "Edit" on the top right of the example, and you'll be taken to a codesandbox with the code of the example.

**Comment by ahocevar** [comment]
> However, as the symptoms are very similar to other issues I have experiensed also with Android and other browsers, the pointer events might not be the one to blame here. 

Those other issues should have been fixed with #11270.

**Comment by t-muko** [comment]
> Those other issues should have been fixed with #11270.

Yes, that was effective fix for the previous issues. Thanks for that!

**Comment by ahocevar** [comment]
@t-muko With iOS 12, can you please try to use a different pointer events polyfill?

Instead of

    <script src="https://unpkg.com/elm-pep"></script>

use

    <script src="https://unpkg.com/@openlayers/pepjs"></script>

**Comment by t-muko** [comment]
Tested with 
<script src="https://unpkg.com/@openlayers/pepjs"></script>   
and
<script src="https://code.jquery.com/pep/0.4.3/pep.js"></script>

No change in behaviour.

**Comment by ahocevar** [comment]
Now what if you go into Safari settings - Experimental features, and enable Pointer Events?

**Comment by t-muko** [comment]
> 
> 
> Now what if you go into Safari settings - Experimental features, and enable Pointer Events?

Ah, good idea.

Enabling pointer events does not help. The problem persists. The only observed change is that now the map gestures work without pep-scripts.

**Comment by t-muko** [comment]
A new observation, if it helps: When Pointer Events was enabled, I managed to get the map to the state where single finger pan was clearly zooming. Just like if I had placed another finger on the bottom left corner of the display. However, after a short while, the "single finger zoom" reset to more usual "slow panning"-mode, which is like the other finger is far away outside of the screen.

**Comment by ahocevar** [comment]
I don't have an iOS 12 device to test, and I cannot reproduce the issue on iOS 13. However, one thing I noticed is that you have not configured the page to work properly on mobile screens. Make sure you add the following meta tag in the head of your markup:
```html
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable: no">
```
Omitting that will cause your device to create map canvases that are bigger than the device can handle.

I have forked your codesandbox with that change: https://codesandbox.io/s/red-fog-2dx5s. Please check if that improves the situation for you.

**Comment by t-muko** [comment]
I have those meta tags in my production code. 

I tested with your sandbox. The problem still persists. Tested with Safari experimental pointer events on.

**Comment by t-muko** [comment]
Just noticed that that viewport meta stuff does not disable the viewport zooming in Safari. Apparently it's by design:
https://stackoverflow.com/questions/4389932/how-do-you-disable-viewport-zooming-on-mobile-safari

**Comment by t-muko** [comment]
For Android devices, your proposed tags are not sufficient either. You need to define
`<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no, maximum-scale=1.0, user-scalable=no" />`

But as said, that does not help for iOS...

**Comment by ahocevar** [comment]
As I said, I cannot reproduce your issue. If you are able to find the cause and provide a pull request, please go for it.

**Comment by t-muko** [comment]
Darn... Any tips where to look for possible culprit? 
Would it be possible to reset the assumed state machine tracking the touch fingers after some timeout or something? 

Somehow the problem seems to be related to the load caused by tile render/source. If I remove the street name layer requiring the heavy processing, the problem goes away on this example. However, I have seen this issue triggered by just another time image map source too. The street name vector layer is just the easiest example to replicate the problem.

**Comment by t-muko** [comment]
I applied the Map pixelRatio: 1 fix to my test code to fix the rendered image scaling (see #11464). After doing that, I can not reproduce the issue with the test code. Tested with VectorTileLayer renderMode: 'hybrid' and renderMode: 'image' even the image rendering takes some.

Could this be some kind of overflow issue after all?

I can still see the issue in my production version even with the pixelRatio: 1 fix. 

However, in that code the problem is triggered by a change from one image tile source to another when the zoom level changes (I have got three layers with XYZ tile sources defined for different zoom levels)

**Comment by t-muko** [comment]
New finding. The transition from OSM layer to XYZ source fails if I define the maxZoom on OSM-layer. 
I don't even need to define another layer:


```
var map = new Map({
    pixelRatio: 1,
    layers: [
        new TileLayer({
            source: new OSM(),
            maxZoom: 14
        })],
    target: 'map',
    view: new View({
        zoom: 14,
    }),
    controls: defaultControls({ zoomOptions: { delta: 0.5 } })
});

```

If I limit the source instead of layer, the pinch zoom crash does not occur!
```
var map = new Map({
    pixelRatio: 1,
    layers: [
        new TileLayer({
            source: new OSM(maxZoom: 14),
        })],
    target: 'map',
    view: new View({
        zoom: 14,
    }),
    controls: defaultControls({ zoomOptions: { delta: 0.5 } })
});


```

**Comment by t-muko** [comment]
As a summary, I have been able to avoid crashes with iOS12 by defining the Map pixelRatio: 1 (fixing the problem for vector tile layer) and not using maxZoom setting for OSM layer, but keeping it rendering behind the other layers but limiting the source to avoid unneccessary map tile loading.
The only side effect is that the OSM Attribution is visible even the map is behind other layers.

**Comment by andrea689** [comment]
I have the same problem using map with 100 markers. New Android devices crash after pan/zoom, while old devices don't.

Using `pixelRatio: 1` solves the problem

**Comment by ahocevar** [comment]
Thanks for sharing your interesting finding about `pixelRatio`, @andrea689

**Comment by ahocevar** [comment]
I'm wondering if this has to do with a fractional pixel ratio. @andrea689, what is the `devicePixelRatio` on the devices where pan/zoom causes a crash?

**Comment by andrea689** [comment]
I have crash with Honor 10 with devicePixelRatio = 3 and Samsung S10 with devicePixelRatio = 3

I don't have crash with Huawei Y6 with devicePixelRatio = 2

**Comment by ahocevar** [comment]
So maybe the issue is that the canvas is bigger than the available memory on the device? What happens if you configure the map with `pixelRatio: 2`?

**Comment by andrea689** [comment]
I tried piexelRatio 2 and no crash occured, if I use piexelRatio 3 I have crash in all devices

**Comment by ZakarFin** [comment]
Just found this issue while debugging our app that crashes constantly on Android Chrome when a vector layer is hidden mid-pinch ("pinch zooming out" so the resolution limit for the layer is hit and the layer is hidden while pinching). Setting `pixelRatio: 1` for map seems to fix this so really appreciate the conversation here :) The `window.devicePixelRatio` for my test device (Samsung Tab A8) reports at 1.5 but at least Samsung S7 and all of the devices that others here have used to test this have the same problem (not sure what the devicePixelRatio is on those). Testing with an iPhone 11 Pro works ok though so it seems to be Android-specific on our case. We are using OpenLayers 6.13.0 (haven't tested with the latest ol).

**Comment by t-muko** [comment]
I am once again seeing this problem. This time it looks like it is triggered by a visibility change of a vectorTileLayer when pich-zooming out with an old iPad. I can also trigger the issue by quickly pressing the zoom out button and then by panning with one finger. 

I have suspected this earlier, but haven't been able to pinpoint it too well. 

Is it possible that touch pointer events are somehow disturbed by the load of other tasks?

**Comment by ahocevar** [comment]
@t-muko This sounds either like #13746, or https://github.com/openlayers/openlayers/issues/13010#issuecomment-979422472. So the first thing to try would be to add the following css:
```css
#map {
  touch-action: none;
}
```

**Comment by t-muko** [comment]
> ```
> #map {
>  touch-action: none;
> }
> ```

Nope. This does not help. It really feels like trackedPointers gets one or two extra pointers, which are never deleted. Is this possible if OL misses a pointer event for some reason? 

```
updateTrackedPointers_(mapBrowserEvent) {
    if (isPointerDraggingEvent(mapBrowserEvent)) {
      const event = mapBrowserEvent.originalEvent;

      const id = event.pointerId.toString();
      if (mapBrowserEvent.type == MapBrowserEventType.POINTERUP) {
        delete this.trackedPointers_[id];
      } else if (mapBrowserEvent.type == MapBrowserEventType.POINTERDOWN) {
        this.trackedPointers_[id] = event;
      } else if (id in this.trackedPointers_) {
        // update only when there was a pointerdown event for this pointer
        this.trackedPointers_[id] = event;
      }
      this.targetPointers = getValues(this.trackedPointers_);
    }
  }
```

**Comment by t-muko** [comment]
Shouldn't there be a handler to clear all tracked pointers if a pointer gets cancelled or lost?
[https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events)

**Comment by ahocevar** [comment]
> crashes consistently on Android Chrome when a vector layer is hidden mid-pinch

This is what's described in #13746. At this point, I have no clue what might be causing this. So any reports or findings would be appreciated. Also if someone has an idea for a fix. @t-muko, it sounds like you might be able to figure out a solution?

**Comment by t-muko** [comment]
Some progress. I added a sanity check to the end of PointerInteraction.prototype.updateTrackedPointers_ on the Pointer.js. 
If there are more than two tracked pointers they are all deleted. This is not a smooth solution yet, but allows to recover after #13746 style touch gesture hangup. And proves the point. Somehow the events are missed on iOS.

```
if (mapBrowserEvent.type == MapBrowserEventType.POINTERCANCEL || Object.keys(this.trackedPointers_).length > 2) {
                console.debug("Deleting all pointer touches")
                for (let i in this.trackedPointers_) 
                    delete this.trackedPointers_[i];
              }
```

**Comment by t-muko** [comment]
ol/MapBrowserEventHandler activePointers_ gets to zero correctly, so events are passed to ol. 
However, 
ol/interaction/Pointer trackedPointers_ count gets increasing when slow operations like visibility change of a vectorTileLayer or some background processing is done during touch actions.

I haven't seen any pointer cancel events, though.

**Comment by Skizu** [comment]
In my case `MapBrowserEventHandler.trackedTouches_` incorrectly has 1 more touch. I am working on a solution which makes use of [TouchEvent.touches](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/touches) to reconcile touches when events are missing.

**Comment by oudoken** [comment]
I can confirm that making a vector layer (in my case geojson or kml but maybe all vector source are affected) with a minimum zoom level "breaks" gestures (pinch/punch,rotation,zoom) on iOS when the vector layer appear.
On android when zooming in with pinch the gesture stop when the vector layer appears.
If the layer is always visible then the problem does not occur.
Not having the time to try but maybe making the layer visible with a timeout on zoom level with map event will work.

**Comment by ahocevar** [comment]
The reason is that on affected platforms the pointer ids won't be reused when the event target changes.

**Comment by ahocevar** [comment]
Please try the potential fix in #13770 and report if it fixes the problem for you.

**Comment by ahocevar** [comment]
I created a better fix in #13771. Please test again.

**Comment by t-muko** [comment]
The problem seems to have resurfaced with iOS 16.4.1(a) hotfix! Touch finger counting messing up. Sometimes tapping with two fingers resumes the normal operation but not always.

**Comment by ahocevar** [comment]
@t-muko I cannot reproduce with the initial description of this issue on an iPhone 13 with iOS 16.4.1(a). Can you provide a reproduction case?

**Comment by t-muko** [comment]
> @t-muko I cannot reproduce with the initial description of this issue on an iPhone 13 with iOS 16.4.1(a). Can you provide a reproduction case?

I am sorry for the slow response. I am still using 6.15 version in my production version. However, I did make a test version using 7.xx release a week ago and asked my iOS testers to try that.
They reported that they can make the two finger gestures crash in similar way as before our last fix.

My testers think that a slow mobile network might make the error more probable to happen. I have got a slightly customised tile loader function to handle some error conditions (server busy, intermittent Internet connection, tile load retry) better than the standard tile loader. I need to run some more tests to verify if there is some relation to that. 

I also noticed that the 7.xx version has got some regression in pre- and post-render animation frame rates. I will investigate that too and raise another issue when I got my head around this.

IF the touch problem is related to some race conditions that are time/load dependent, these observations might make sense.
